### PR TITLE
Week03

### PR DESCRIPTION
**Pull request by safiweb**
# Adding two  struct types, methods to struct type, a new function type, and test functions

This PR creates:

### - Two structs _Movie_ and _Theater_

**Movie** - It export two public fields; Length of type _int_ and Name of type _string_.
**Theater** -  It does not export any public fields.

### - A new function type named CritiqueFn
It creates a new type named _CritiqueFn_ that is a function type that takes a pointer of type _Movie_ and returns a _float32_ and an _error_.

### - Six methods defined on struct Movie type

> _**_Rate_**_
> Rate is a method on the pointer receiver for Movie. Rate take a _float32_ rating and return an _error_. 
> 
> _**Play**_
> Play is a method on the pointer receiver for Movie. Play take the number (_int_) of viewers watching the movie. Calling Play increases both the number of viewers, as well as the number of plays, for the movie.
> 
> _**Viewers**_
> Viewers is a method on the value receiver for Movie. Viewers take no arguments and return the number (_int_) of people who have viewed the movie.
> 
> _**Plays**_
> Plays is a method on the value receiver for Movie. Plays take no arguments and return the number (int) of times the movie has been played.
> 
> _**Rating**_
> Rating is a method on the value receiver for Movie. Rating takes no arguments and returns the rating (_float64_) of the movie. This is calculated by the total ratings for the movie divided by the number of times the movie has been played.
> 
> _**String**_
> String is a method on the value receiver for Movie. String returns a _string_ that includes the name, length, and rating of the film. Ex. Wizard of Oz (102m) 99.0%

### - Two methods defined on struct Theater type

> **_Play_**
> Play is a method on the pointer receiver for Theatre. Play take the number (_int_) of viewers, a variadic list of pointer of type Movie, and return an _error_. Calling Play will call each movie’s Play method with the number of viewers. If no movies are passed it return the following error: _fmt.Errorf("no movies to play")_.
> 
> **_Critique_**
> Critique is a method on the value receiver for Theatre. Critique should take a slice of pointers to type Movie, a _CritiqueFn_, and return an _error_. Calling Critique will iterate over the movies and for each movie. First, each movie’s Play method is called with a value of 1. Next, the CritiqueFn is called with the movie, it return values that are error checked. If there is no error the movie’s Rate method is called with the float32 value that was returned from the CritiqueFn. 

All the above are added via assignment02.go.

### - Test functions for all methods and functions defined
The test functions are added via assignment03.test.go

### TestRate
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie and output are of type error. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestRate ./...
=== RUN   TestRate
=== RUN   TestRate/#00
=== RUN   TestRate/#01
    assignment03_test.go:23: no movies to play: can't review a movie without watching it first
--- FAIL: TestRate (0.00s)
    --- PASS: TestRate/#00 (0.00s)
    --- FAIL: TestRate/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 1.194s
FAIL
```

### TestPlay
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie and int, the output is of type int. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestPlay ./...
=== RUN   TestPlay
=== RUN   TestPlay/#00
=== RUN   TestPlay/#01
    assignment03_test.go:44: the Léon: The Professional movie views 5, expected: 0.
    assignment03_test.go:47: the Léon: The Professional movie plays 5, expected: 0.
--- FAIL: TestPlay (0.00s)
    --- PASS: TestPlay/#00 (0.00s)
    --- FAIL: TestPlay/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.955s
FAIL
```

### TestViewers
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie, the output is of type error. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestViewers ./...
=== RUN   TestViewers
=== RUN   TestViewers/#00
    assignment03_test.go:66: the Gladiator movie viewers -1, is not a number.
=== RUN   TestViewers/#01
--- FAIL: TestViewers (0.00s)
    --- FAIL: TestViewers/#00 (0.00s)
    --- PASS: TestViewers/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 1.429s
FAIL
```

### TestPlays
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie, the output is of type error. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestPlays ./...
=== RUN   TestPlays
=== RUN   TestPlays/#00
    assignment03_test.go:85: the Gladiator movie plays -1, is not a number.
=== RUN   TestPlays/#01
--- FAIL: TestPlays (0.00s)
    --- FAIL: TestPlays/#00 (0.00s)
    --- PASS: TestPlays/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.654s
FAIL
```

### TestRating
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie, the output is of type error. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestRating ./...
=== RUN   TestRating
=== RUN   TestRating/#00
=== RUN   TestRating/#01
    assignment03_test.go:104: the Léon: The Professional movie rating 0.0 should be greater than 0, 0 is not a rating number.
--- FAIL: TestRating (0.00s)
    --- PASS: TestRating/#00 (0.00s)
    --- FAIL: TestRating/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.753s
FAIL
```

#### TestString
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Movie and string, the output is of type error. It then tests for success or failure.
```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestString ./...
=== RUN   TestString
=== RUN   TestString/#00
=== RUN   TestString/#01
    assignment03_test.go:124: error outputing movie detail,got: Léon: The Professional (110m) 4.5%, expected:Léon: The Professional (110m) 68.9%
--- FAIL: TestString (0.00s)
    --- PASS: TestString/#00 (0.00s)
    --- FAIL: TestString/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.160s
FAIL
```

#### TestPlayTheater
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Theater, Movie and error, the output is of type error. It then tests for success or failure.
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestPlayTheater ./...
```
=== RUN   TestPlayTheater
=== RUN   TestPlayTheater/#00
=== RUN   TestPlayTheater/#01
    assignment03_test.go:144: no movies to play
--- FAIL: TestPlayTheater (0.00s)
    --- PASS: TestPlayTheater/#00 (0.00s)
    --- FAIL: TestPlayTheater/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 1.310s
FAIL
```

#### TestCritique
It set a tests struct type that holds our test inputs and expected outputs. The inputs are of type Theater, Movie and error, the output is of type error. It then tests for success or failure.

E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestCritique ./...
```
=== RUN   TestCritique
=== RUN   TestCritique/#00
    assignment03_test.go:181: no rate provided or exist
    assignment03_test.go:181: no rate provided or exist
=== RUN   TestCritique/#01
    assignment03_test.go:181: no rate provided or exist
    assignment03_test.go:181: no rate provided or exist
=== RUN   TestCritique/#02
--- FAIL: TestCritique (0.00s)
    --- FAIL: TestCritique/#00 (0.00s)
    --- FAIL: TestCritique/#01 (0.00s)
    --- PASS: TestCritique/#02 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.742s
FAIL
```
## Changes Summary
Added new file assignment03.go, the main file with structs and methods definitions
Added new file assignment03_test.go, the unit test file
Added Rate, Play, Viewers, Plays, Rating, String methods to Movie type
Added CritiqueFn function type
Added Play, Critique. methods to Theater type
Added TestRate, TestPlay, TestViewers, TestPlays, TestRating, TestString, TestPlayTheater, TestCritique test functions
Miscellaneous - formatting code

## Difficulties:
1. Slices of structs vs. slices of pointers to structs, which one to use e.g using the struct Movie type:
`[]*Movie or []Movie`

2. Running unit test function TestPlay alone also runs TestPlays and TestPlayTheater e.g

```
E:\Projects\Go\src\gopherguides-intro-to-go\week03>go test -v -run TestPlay ./...
=== RUN   TestPlay
=== RUN   TestPlay/#00
=== RUN   TestPlay/#01
    assignment03_test.go:44: the Léon: The Professional movie views 5, expected: 0.
    assignment03_test.go:47: the Léon: The Professional movie plays 5, expected: 0.
--- FAIL: TestPlay (0.00s)
    --- PASS: TestPlay/#00 (0.00s)
    --- FAIL: TestPlay/#01 (0.00s)
=== RUN   TestPlays
=== RUN   TestPlays/#00
    assignment03_test.go:85: the Gladiator movie plays -1, is not a playable number.
=== RUN   TestPlays/#01
--- FAIL: TestPlays (0.00s)
    --- FAIL: TestPlays/#00 (0.00s)
    --- PASS: TestPlays/#01 (0.00s)
=== RUN   TestPlayTheater
=== RUN   TestPlayTheater/#00
=== RUN   TestPlayTheater/#01
    assignment03_test.go:144: no movies to play
--- FAIL: TestPlayTheater (0.00s)
    --- PASS: TestPlayTheater/#00 (0.00s)
    --- FAIL: TestPlayTheater/#01 (0.00s)
FAIL
FAIL    gopherguides-intro-to-go/week03 0.148s
FAIL
```
3. How to check if a struct has fields or methods i.e if it's undefined or empty. For example for struct Movie type , I was not able to do something like

```
v : = Movie{}
if(v== nil){
  fmt.Println("Movie struct is undefined/fields don't exist")
}
```
4. The best way to define test tables and make use of t.Run() & t.Parallel() for testing success and failure scenarios.
5. The best Go way to arrange a project,  got stuck separating my work into their own packages. I was forced to revert to the easier way. For example, this is what I had arranged for my work:
image

## Surprises
1. Structs, methods defined for the structs, exported/non-exported struct fields; the concept is similar to OO in other languages. In Go, it's a bit cleaner.
2. Using test tables for testing different user scenarios. I found it cool instead of writing long code to test each scenario.
- [https://dave.cheney.net/2019/05/07/prefer-table-driven-tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)
- [https://www.gopherguides.com/articles/table-driven-testing-in-parallel](https://www.gopherguides.com/articles/table-driven-testing-in-parallel)
3. When defining structs and their methods on a file calling package main, function main can be empty i.e
```
func main(){
}
```
4. Discovered I could use TypeOf  method from package reflect to check if variables are of the same type.

### CODE REVIEW

File: week03/assignment03.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+type Theater struct{}
+
+func (m *Movie) Rate(rating float32) error {
+
+	if m.plays == 0 {
+		return fmt.Errorf("can't review a movie without watching it first")
+	}
+
+	m.rating = m.rating + rating
```
Comment (markbates) [inline week03/assignment03.go]:
```suggestion
	m.rating += rating
```

File: week03/assignment03.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	return nil
+
+}
+
+func (m *Movie) Play(viewers int) {
+	m.viewers = m.viewers + viewers
+	m.plays = m.plays + viewers
```
Comment (markbates) [inline week03/assignment03.go]:
```suggestion
	m.viewers += viewers
	m.plays++
```

File: week03/assignment03.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	ratings := m.rating / float32(m.plays)
+	return float64(ratings)
+}
+
+func (m Movie) String() string {
+	s := fmt.Sprintf("%v (%dm) %.1f%%", m.Name, m.Length, m.rating)
+	return s
+}
+
+func (t *Theater) Play(viewers int, m ...*Movie) error {
```
Comment (markbates) [inline week03/assignment03.go]: You want to be able to use the variable `m` when referring to a single movie in your code. Use something like `movies` for the slice of movies you capture. It's clearer and frees up `m` as a variable name.

```suggestion
func (t *Theater) Play(viewers int, movies ...*Movie) error {
```

File: week03/assignment03.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	for _, v := range m {
+		if v.Name == "" && v.Length == 0 {
+			return fmt.Errorf("no movies to play")
+		}
```
Comment (markbates) [inline week03/assignment03.go]: This checks to see if that particular movie is "empty". This check will not be called if the list of movies itself is empty. That is where the check should be.

File: week03/assignment03.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+func (m Movie) String() string {
+	s := fmt.Sprintf("%v (%dm) %.1f%%", m.Name, m.Length, m.rating)
+	return s
+}
+
+func (t *Theater) Play(viewers int, m ...*Movie) error {
+
+	for _, v := range m {
```
Comment (markbates) [inline week03/assignment03.go]:
```suggestion
	if len(movies) == 0 {
		return fmt.Errorf("no movies to play")
	}
	
	for _, m := range movies {
```

File: week03/assignment03.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+func (t *Theater) Critique(fn CritiqueFn, m []*Movie) error {
+	for _, v := range m {
+		v.Play(1)
+		if rate, err := fn(v); err != nil {
+			return fmt.Errorf("no rate provided or exist")
+		} else {
+			if err := v.Rate(rate); err != nil {
+				return fmt.Errorf("oh no, something went wrong! %v", err)
+			}
+		}
```
Comment (markbates) [inline week03/assignment03.go]: This set of clauses is far too complex. The logic is hard to follow. In Go, we try to eliminate the need for complex statements like these by trying to not use the `else` clause. 

```go
func (t *Theatre) Critique(fn CritiqueFn, m []*Movie) error {
	for _, v := range m {
		v.Play(1)

		rate, err := fn(v)
		if err != nil {
			return fmt.Errorf("no rate provided or exist")
		}

		if err := v.Rate(rate); err != nil {
			return fmt.Errorf("oh no, something went wrong! %w", err)
		}
	}

	return nil
}
```

File: week03/assignment03_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func TestRate(t *testing.T) {
+
+	var tests = []struct {
```
Comment (markbates) [inline week03/assignment03_test.go:10]: We will cover table driven tests next week, but good use of them here.

File: week03/assignment03_test.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		want error
+	}{
+		{Movie{Name: "Gladiator", Length: 155, plays: 4}, nil},
+		{Movie{Name: "Léon: The Professional", Length: 110}, fmt.Errorf("can't review a movie without watching it first")},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			got := tt.a.Rate(50.2)
+			if got != tt.want {
```
Comment (markbates) [inline week03/assignment03_test.go:22]: As we will learn next week, we want to be careful when checking errors this exactly. Instead, it is considered idiomatic to confirm whether an error is `nil` or not.

File: week03/assignment03_test.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	var tests = []struct {
+		a    Movie
+		want error
+	}{
+		{Movie{Name: "Gladiator", Length: 155, plays: 4}, nil},
+		{Movie{Name: "Léon: The Professional", Length: 110}, fmt.Errorf("can't review a movie without watching it first")},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
```
Comment (markbates) [inline week03/assignment03_test.go]: Always name your subtests. We'll learn why next week. :)

```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

File: week03/assignment03_test.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
```
Comment (markbates) [inline week03/assignment03_test.go]: There is no reason for these tests to be using the `reflect` package. The use of `reflect` is considered both "advanced" and, often, ill-advised. Use of reflection in code needs to be justified. The requirements for these tests don't justify reflection.

File: week03/assignment03_test.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+		a Movie
+	}{
+		{Movie{Name: "Gladiator", Length: 155, viewers: -1}},
+		{Movie{Name: "Léon: The Professional", Length: 110, viewers: 4}},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			got := tt.a.Viewers()
+			if reflect.TypeOf(got) != reflect.TypeOf(0) || got < 0 {
```
Comment (markbates) [inline week03/assignment03_test.go]: There is no need for use of reflection here. `got` is a concrete type, `int`. All this code does is confirm what the compiler is already doing. 
```suggestion
			if got < 0 {
```

File: week03/assignment03_test.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		a Movie
+	}{
+		{Movie{Name: "Gladiator", Length: 155, plays: -1}},
+		{Movie{Name: "Léon: The Professional", Length: 110, plays: 4}},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			got := tt.a.Plays()
+			if reflect.TypeOf(got) != reflect.TypeOf(0) || got < 0 {
```
Comment (markbates) [inline week03/assignment03_test.go]: Again, no use of reflection necessary.

File: week03/assignment03_test.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+	}{
+		{Theater{}, Movie{Name: "Gladiator", Length: 155}, nil},
+		{Theater{}, Movie{}, fmt.Errorf("no movies to play")},
+	}
+
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			if err := tt.a.Play(15, &tt.b); err != nil {
+				t.Errorf("no movies to play")
+			}
```
Comment (markbates) [inline week03/assignment03_test.go]: This is not testing what you think it is. This only errors if the `Play` method returns an error. As long as `Play` does not error, this test will always succeed. Even if the test succeeds you are not making any assertions that the call to `Play` did what it should have.  In addition, if the test does fail, you don't supply the actual error to `t.Errorf`, but rather your own error message. Re-consider what the objective of this test is. It should test positive and negative scenarios.

File: week03/assignment03_test.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+				v.Play(1)
+				//fn(v)
+				critiqueRating -= 2
+				if rate, err := fn(v); err != nil || rate <= 0 {
+					t.Errorf("no rate provided or exist")
+				} else {
+					if err := v.Rate(rate); err != nil {
+						t.Errorf("oh no, something went wrong! %v", err)
+					}
+				}
```
Comment (markbates) [inline week03/assignment03_test.go]: Again try cleaning up this set of `if/else` statements.

**Review by markbates — COMMENTED** [review]
Overall good work. Please make changes and resubmit by Tuesday for full marks.

File: week03/assignment03.go (hunk 1 above)
Comment (safiweb) [reply week03/assignment03.go]: Done

File: week03/assignment03.go (hunk 2 above)
Comment (safiweb) [reply week03/assignment03.go]: Done

File: week03/assignment03_test.go (hunk 9 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok, cool. I was confused about naming the subtests but it all cleared up :)

File: week03/assignment03_test.go (hunk 10 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok, noted. I refactored the code and removed the usage of reflect package.

File: week03/assignment03_test.go (hunk 11 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok

File: week03/assignment03_test.go (hunk 12 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok, noted.

File: week03/assignment03_test.go (hunk 13 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok, I had thought I don't need to test a function/method called inside another function/method. After your explanation, all cleared up.

File: week03/assignment03_test.go (hunk 14 above)
Comment (safiweb) [reply week03/assignment03_test.go]: ok, I have refactored the if/else statements.

File: week03/assignment03_test.go (hunk 8 above)
Comment (safiweb) [reply week03/assignment03_test.go:22]: ok

File: week03/assignment03_test.go (hunk 7 above)
Comment (safiweb) [reply week03/assignment03_test.go:10]: cool :)

File: week03/assignment03.go (hunk 6 above)
Comment (safiweb) [reply week03/assignment03.go]: cool, it's the PHP in me. I will stay away else clauses. I have refactored the code.

File: week03/assignment03.go (hunk 5 above)
Comment (safiweb) [reply week03/assignment03.go]: ok

File: week03/assignment03.go (hunk 4 above)
Comment (safiweb) [reply week03/assignment03.go]: ok

File: week03/assignment03.go (hunk 3 above)
Comment (safiweb) [reply week03/assignment03.go]: ok, cool.
